0013-RS-pubsub-subscription-streaming.md

```diff
@@ -0,0 +1,159 @@
+
+Today, it is not possible to dynamically update the subscription list without restarting Daprd, though hot reloading for Subscription manifests is [planned](https://github.com/dapr/dapr/issues/7139).
+It is common for users to want to dynamically subscribe and unsubscribe to topics inside their applications based on runtime conditions.
+In the cases where Dapr is not running as a sidecar, users often do not want to open a public port or create a tunnel in order to receive PubSub messages from Dapr.
+
+A streaming Subscription API will allow applications to dynamically subscribe to PubSub topics and receive messages without opening a port to receive incoming traffic from Dapr.
+
+## Expectations and alternatives
+
+This proposal outlines the gRPC & HTTP streaming API for subscribing to PubSub topics.
+This proposal does _not_ address any hot-reloading functionality to the existing programmatic or declarative subscription configuration.
+Using a gRPC streaming API is the most natural fit for this feature, as it allows for first class long-lived bi-directional connections to Dapr to receive messages.
+A supplementary WebSocket based HTTP API is useful for applications which do not have a gRPC client available or HTTP WebSockets are preferred.
+These messages are typed RPC giving the best UX in each SDK.
+Once implemented, this feature will need to be implemented in all Dapr SDKs.
+
+## Solution
+
+### gRPC
+
+Rough gRPC PoC implementation: https://github.com/dapr/dapr/commit/ed40c95d11b78ab9a36a4a8f755cf89336ae5a05
+
+The Dapr runtime gRPC server will implement the following new RPC and messages:
+
+```proto
+service Dapr {
+  // SubscribeTopicEvents subscribes to a PubSub topic and receives topic events
+  // from it.
+  rpc SubscribeTopicEvents(stream SubscribeTopicEventsRequest) returns (stream TopicEventRequest) {}
+}
+
+// SubscribeTopicEventsRequest is a message containing the details for
+// subscribing to a topic via streaming.
+// The first message must always be the initial request. All subsequent
+// messages must be event responses.
+message SubscribeTopicEventsRequest {
+  oneof subscribe_topic_events_request_type {
+    SubscribeTopicEventsSubscribeRequest request = 1;
+    SubscribeTopicEventsResponse event_response = 2;
+  }
+}
+
+// SubscribeTopicEventsSubscribeRequest is the initial message containing the
+// details for subscribing to a topic via streaming.
+message SubscribeTopicEventsSubscribeRequest {
+  // The name of the pubsub component
+  string pubsub_name = 1 [json_name = "pubsubName"];
+
+  // The pubsub topic
+  string topic = 2 [json_name = "topic"];
+
+  // The metadata passing to pub components
+  //
+  // metadata property:
+  // - key : the key of the message.
+  map<string, string> metadata = 3 [json_name = "metadata"];
+
+  // dead_letter_topic is the topic to which messages that fail to be processed
+  // are sent.
+  optional string dead_letter_topic = 4 [json_name = "deadLetterTopic"];
+}
```

Comment:
can we add a `max_in_flight_messages` and `max_await_duration_ms` similar to bulk subscribe API https://docs.dapr.io/developing-applications/building-blocks/pubsub/pubsub-bulk/#subscribing-messages-in-bulk

But in this case it is the max number of in-flight messages that we allow before waiting for an acknowledgement or max await time before we wait for an acknowledgement.

Reply:
Added `max_in_flight_messages`, however `max_await_duration_ms` would already be covered by existing resiliency policy.